From 769f1d95e2c3770c6bd2829d88802b02b18aefd7 Mon Sep 17 00:00:00 2001
From: tnviet <ngocviet.tran1987@gmail.com>
Date: Mon, 21 Jun 2021 12:27:02 +0700
Subject: [PATCH 131/196] Allow seeder to run when there are no checkpoints
 (#199)

Currently, the seeder segfaults without any checkpoints. This commit
adjust the lookups into mapCheckpoints to first see if it has any elements.
This allows setting up Lotus seeders since we do not have any initial
checkpoints.

Co-authored-by: vince8x <vince.tran8x@gmail.com>
Co-authored-by: Shammah Chancellor <shammah.chancellor@gmail.com>
---
 src/seeder/bitcoin.cpp | 6 ++++--
 src/seeder/db.h        | 6 +++++-
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/src/seeder/bitcoin.cpp b/src/seeder/bitcoin.cpp
index 5cf710f3a..f20c48b84 100644
--- a/src/seeder/bitcoin.cpp
+++ b/src/seeder/bitcoin.cpp
@@ -61,8 +61,10 @@ PeerMessagingState CSeederNode::ProcessMessage(std::string strCommand,
         // nVersion);
         if (vAddr) {
             MessageWriter::WriteMessage(vSend, NetMsgType::GETADDR);
-            std::vector<BlockHash> locatorHash(
-                1, Params().Checkpoints().mapCheckpoints.rbegin()->second);
+            std::vector<BlockHash> locatorHash(0);
+            if (Params().Checkpoints().mapCheckpoints.size() > 0) {
+                locatorHash = std::vector<BlockHash>(1, Params().Checkpoints().mapCheckpoints.rbegin()->second);
+            }
             MessageWriter::WriteMessage(vSend, NetMsgType::GETHEADERS,
                                         CBlockLocator(locatorHash), uint256());
             doneAfter = GetTime() + GetTimeout();
diff --git a/src/seeder/db.h b/src/seeder/db.h
index c84a7c903..a339f3b8f 100644
--- a/src/seeder/db.h
+++ b/src/seeder/db.h
@@ -26,7 +26,11 @@
 #define REQUIRE_VERSION 70001
 
 static inline int GetRequireHeight() {
-    return Params().Checkpoints().mapCheckpoints.rbegin()->first;
+    if (Params().Checkpoints().mapCheckpoints.rbegin() !=
+        Params().Checkpoints().mapCheckpoints.rend()) {
+        return Params().Checkpoints().mapCheckpoints.rbegin()->first;
+    }
+    return 0;
 }
 
 static inline std::string ToString(const CService &ip) {
-- 
2.39.5 (Apple Git-154)

